### DISCUSSION TITLE
filter(Predicate) vs. takeWhile(Predicate) or dropWhile(Predicate)

### DISCUSSION THREAD

**Discussion by corock** [Study-2-Modern-Java-In-Action/Study-2-Modern-Java-in-Action#51] [CHAPTER 5 스트림 활용 - Q&A] [answered]
> _p. 158_ 내용에 의하면... 자바 9부터 스트림의 요소를 효과적으로 선택할 수 있도록 `takeWhile`, `dropWhile` 메서드를 지원한다.
>
> 기존에 프리디케이트를 이용한 슬라이싱으로 사용했던 `filter` 와의 차이점으로 전체 탐색 대신 조건 만족 시 부분 탐색이 가능한데,
> 그렇다면 성능 상 무조건  `filter` 메서드보다 `takeWhile`, `dropWhile` 메서드를 사용하는 것이 좋은 것이 아닌가?

**Comment by coalong** [comment, marked as answer]
제 생각은 정렬되어 있는 요소를 탐색할 때는 
성능상 `filter` 를 사용하는 것보다 `takeWhile`, `dropWhile` 을 사용하는 것이 좋다고 생각합니다.

 만약 정렬되어 있지 않은 요소를 탐색할 때는 `takeWhile` 이나 `dropWhile` 로 원하는 결과를 못 얻을 수 있다고 생각되는데, 
이런 상황에서는 1) `sorted` 와 `takeWhile` or `dropWhile` 를 같이 사용하는 경우, 
2)`filter` 을 사용하는 경우 어떤 것이 더 좋을지 궁금합니다.

**Reply by BvrPark** [reply]
저도 이게 궁금합니다!
아무리 찾아봐도 `takeWhile`, `dropWhile`메서드는 최악의 상황이어야지 전체 조건을 탐색하는 `filter`랑 성능이 비슷하다고 나오는데
그러면 무조건 `filter`의 성능이 `takeWhile`, `dropWhile`에 비해서 성능이 떨어지는 것이 아닌지 궁금합니다.

**Reply by JoisFe** [reply]
`takeWhile, dropWhile 최악의 상황 경우 성능이 같고 그 외에는 무조건 성능이 떨어진다`
- 정렬된 상태라 가정하면 filter가 takeWhile, dropWhile이 최악의 상황인 경우 성능이 같고 그 외에는 성능이 더 안 좋은 것은 맞습니다. 
(최악의 상황 : 스트림의 데이터들 무게가 모두 500kg 이하이고 오름차순 정렬이 되어있는 상황에서 600kg 짜리를 찾을려고 하는 상황이 아닐까 싶습니다. -> filter든 takeWhile, droptWhile 결국 모두 모든 데이터를 다 탐색 해야하니깐요)
- Filter 보다 takeWhile, dropWhile 성능이 좋다면 무조건 filter 말고 정렬 후 dropWhile, takeWhile을 사용하면 되는 것 아닌가?
  - 만약 슬라이싱을 한번만 진행한다면 굳이 시간복잡도가 더 높은 정렬을 하는게 오히려 더 손해이기 때문에 상황에 따라 다른 것 같습니다.
